src/graph/impl/KokkosGraph_ExplicitCoarsening_impl.hpp

```diff
@@ -0,0 +1,303 @@
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOSGRAPH_EXPLICIT_COARSEN_IMPL_HPP
+#define KOKKOSGRAPH_EXPLICIT_COARSEN_IMPL_HPP
+
+namespace KokkosGraph {
+namespace Impl {
+
+template<typename lno_t, typename size_type, typename device_t, typename fine_rowmap_t, typename fine_entries_t, typename labels_t, typename coarse_rowmap_t, typename coarse_entries_t, typename ordinal_view_t>
+struct ExplicitGraphCoarsening
+{
+  using exec_space = typename device_t::execution_space;
+  using range_pol = Kokkos::RangePolicy<exec_space>;
+  using team_pol = Kokkos::TeamPolicy<exec_space>;
+  using team_member_t = typename team_pol::member_type;
+  using bitset_t = Kokkos::Bitset<device_t>;
+  using const_bitset_t = Kokkos::ConstBitset<device_t>;
+
+  struct ClusterSizeFunctor
+  {
+    ClusterSizeFunctor(const ordinal_view_t& counts_, const labels_t& vertClusters_)
+      : counts(counts_), vertClusters(vertClusters_)
+    {}
+    KOKKOS_INLINE_FUNCTION void operator()(const lno_t i) const
+    {
+      Kokkos::atomic_increment(&counts(vertClusters(i)));
```

Comment:
Why not have `counts` use the atomic memory trait?

Reply:
@lucbv I guess it could but in exchange for making the increment line simpler, I would have to add another typedef for ``atomic_ordinal_view_t``, so I think it comes out even